### PR TITLE
An implementation of id_print with CustomCall

### PR DESCRIPTION
**Pull request by gnecula**
The implementation uses CustomCall and works only on CPU for now.

 Behaves like identify function for positional arguments, but prints all arguments.

 The return value is a tuple with the value of `args` or the value of the keyword
  parameter `result` if present. If there is a single positional argument, returns
  just that argument without packing it in a tuple.

  The positional arguments must be JAX values. The keyword arguments are serialized
  to a string and printed along with the positional arguments. The exception is the
  `result` keyword argument, which must be a JAX value.

     Usage:
     >>> y = id_print(x * 2)  # prints and returns 2x
     >>> y, z = id_print(x * 2, x * 3)  # prints and returns 2x and 3x
     >>> y = id_print(x * 2, result=y)  # prints 2x and returns y
     >>> y = id_print(x * 2, what='x')  # prints what=x followed by 2x

  The order of execution is by data dependency: after all the arguments are computed
  and before the result is used. At least one of the returned values must be used
  in the rest of the computation, or else this operation has no effect.

Upon JAX transformations, the transformed values are wrapped with `id_print`, 
and a special `transforms` tuple keyword argument is added with the sequence of
transformations applied:

   - For `vmap` the arguments are batched, and transforms=('vmap')
   - For `jvp` there will be an id_print for the primal values, and a separate
      `id_print` for the tangents with `transforms=('jvp')`.
   - For `grad` there will be an `id_print` for the primal values (if needed
          in the computation of `grad` and an `id_print` with the adjoints of the
          results, with transforms=('vjp').

### CODE REVIEW

File: jax/experimental/host_callback.py [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+from jax.interpreters import partial_eval as pe
+from jaxlib import xla_client
+from jaxlib import xla_extension
+
+import msgpack  # type: ignore
+import numpy as onp
+import os
+from typing import Any, Dict, List, Sequence, Tuple
+
+if os.getenv("JAX_ENABLE_JIT_PRINT", "false") != "false":
```
Comment (gnecula) [inline jax/experimental/host_callback.py:54]: This is a temporary flag to allow Travis to test the Python code, even before the C++ code made it into jaxlib.

File: jax/experimental/host_callback.py [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+# TODO(necula): fix mypy errors if I define the type aliases below
+XlaOp = Any  # xla_extension.XlaOp
+XlaShape = Any # xla_client.Shape 
+XlaComputationBuilder = Any  # xla_bridge._JaxComputationBuilder
+
+id_print_p = core.Primitive("id_print")
+id_print_p.multiple_results = True
+
+
+def id_print(*args, **kwargs):
```
Comment (shoyer) [inline jax/experimental/host_callback.py:73]: In Python 3, you can use a real keyword argument between `*args` and `**kwargs`:
```suggestion
def id_print(*args, result=None, **kwargs):
```

File: jax/experimental/host_callback.py [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+  """Behaves like the identify function for positional arguments, but prints all
+     arguments on the host, even from transformed or compiled code.
+
+     The return value is a tuple with the value of `args` or the value of the
+     keyword parameter `result` if present. If there is a single positional
+     argument, it returns just that argument without packing it in a tuple.
+
+     The positional arguments must be JAX values. The keyword arguments are
+     serialized to a string and printed along with the positional arguments.
```
Comment (shoyer) [inline jax/experimental/host_callback.py:82]: Would it be feasible to support strings inline in `*args` instead? The rule might be that strings are dropped from `args` in the default return value, e.g,. so you can write something like `x = id_print('x =', x)` or `x, y = id_print('x =', x, 'y =', y)`.

Using the keyword argument for extra strings is a little more awkward to write, and it isn't 100% clear that only strings values are OK. I would sort of expect `id_print(what=x)` to print `what=` followed by the value of `x`.

**Review by shoyer — COMMENTED** [review]
I'm very excited about this, should make a big difference for debugging!

File: jax/experimental/host_callback.py (hunk 2 above)
Comment (gnecula) [reply jax/experimental/host_callback.py:73]: Yes, I'll do this.

File: jax/experimental/host_callback.py (hunk 3 above)
Comment (gnecula) [reply jax/experimental/host_callback.py:82]: These are all good suggestions, I will play with them, I suspect that much of this can be added to the API.

**Comment by hawkinsp** [comment]
I assume this PR is in a draft state and the intent isn't to merge it as is? (If so, mark it as draft?)

**Comment by gnecula** [comment]
Closing, we have now an implementation using outfeed.

**Pull request closed without merging**